### PR TITLE
Layer selection

### PR DESCRIPTION
**Pull request by akshay1992kalbhor**
Very crude version. This is my first time writing JS seriously

---
This change is [Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/166)

### CODE REVIEW

**Review by Keavon — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/166)* status: 0 of 3 files reviewed, 5 unresolved discussions (waiting on @akshay1992kalbhor)

___
*[client/web/.DS_Store, line 0 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-MbZyGS4DOzD2ugIfNhb:-MbZyGS4DOzD2ugIfNhc:b-n6d6dn) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/.DS_Store#L0)):*
Please remove
___
*[client/web/src/components/panels/LayerTree.vue, line 21 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-MbZyRjCCRO64hesvqT9:-MbZyRjCCRO64hesvqTA:brhu3h4) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/src/components/panels/LayerTree.vue#L21)):*
> ```Vue
> 					<div
> 						class="layer"
> 						:class="layer.selected ? 'selected' : ''"
> ```

For purely stylistic reasons, could you please replace this with:
```
:class="{selected: layer.selected}"
```
___
*[client/web/src/components/panels/LayerTree.vue, line 135 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-Mb_4IN4BCC9R6uvLqQJ:-Mb_4IN4BCC9R6uvLqQK:b-tnwq18) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/src/components/panels/LayerTree.vue#L135)):*
> ```Vue
> 								break;
> 							}
> 							j += 1;
> ```

`j++;`
___
*[client/web/src/components/panels/LayerTree.vue, line 155 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-Mb_4PPI0fO3X5KT6BmM:-Mb_4PPI0fO3X5KT6BmN:b-togiml) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/src/components/panels/LayerTree.vue#L155)):*
> ```Vue
> 					}
> 				}
> 				i += 1;
> ```

`i++;`
___
*[client/web/src/components/panels/LayerTree.vue, line 210 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-Mb_-syQ1f7yFxY0nCgR:-Mb_-syQ1f7yFxY0nCgS:bgg6gn3) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/src/components/panels/LayerTree.vue#L210)):*
> ```Vue
> 			SeparatorType,
> 			layers: [] as Array<LayerPanelEntry>,
> 			startPath: -1n,
> ```

A clearer name might be `selectionRangeStartLayer` and `selectionRangeEndLayer`. It should probably be of type `LayerPanelEntry | undefined` (so in TypeScript I believe this line would say `selectionRangeStartLayer: undefined as LayerPanelEntry | undefined,`). You will store a *reference* to the layer (not only the path) that's also in the `layers` array (see the line above this comment). Then for the range selection, the algorithm can find where the `selectionRangeStartLayer` and `selectionRangeEndLayer` are within the array (and eventually within the tree once we have folders) using  `[a, b, c].indexOf(b)`.

I will leave more detailed feedback on your three click selection functions once you switch over the data type.

Your JavaScript/TypeScript is looking seriously awesome despite your minimal experience using it!

File: client/web/src/components/panels/LayerTree.vue [hunk 1]
```diff
@@ -20,7 +20,13 @@
 					<div class="layer-visibility">
 						<IconButton :icon="layer.visible ? 'EyeVisible' : 'EyeHidden'" @click="toggleLayerVisibility(layer.path)" :size="24" :title="layer.visible ? 'Visible' : 'Hidden'" />
 					</div>
-					<div class="layer">
+					<div
+						class="layer"
+						:class="{ selected: layer.layer_data.selected }"
+						@click.shift.exact="handleShiftClick(layer)"
+						@click.alt.exact="handleControlClick(layer)"
```
Comment (Keavon) [inline client/web/src/components/panels/LayerTree.vue:27]: Add another line:
```
						@click.ctrl.exact="handleControlClick(layer)"
```

Someday we'll properly deal with platform-specific keybinding differences with Mac.

File: client/web/src/components/panels/LayerTree.vue [hunk 2]
```diff
@@ -68,6 +74,9 @@
 			margin-left: 4px;
 			padding-left: 16px;
 		}
+		.selected {
+			background: var(--color-accent);
```
Comment (Keavon) [inline client/web/src/components/panels/LayerTree.vue:78]: Please also add a line with `color: var(--color-f-white);`

**Review by TrueDoctor — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/166)* status: 0 of 3 files reviewed, 7 unresolved discussions (waiting on @akshay1992kalbhor and @Keavon)

___
*[client/web/.DS_Store, line  at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-MbZyGS4DOzD2ugIfNhb:-MblA9RU3srbqbP8FlKB:bs7wumb) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/.DS_Store#L)):*
<details><summary><i>Previously, Keavon (Keavon Chambers) wrote…</i></summary><blockquote>

Please remove
</blockquote></details>

Done.
___
*[client/web/src/components/panels/LayerTree.vue, line 21 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-MbZyRjCCRO64hesvqT9:-Mbl9qb41iL-V-XK-RLN:b-896fix) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/src/components/panels/LayerTree.vue#L21)):*
<details><summary><i>Previously, Keavon (Keavon Chambers) wrote…</i></summary><blockquote>

For purely stylistic reasons, could you please replace this with:
```
:class="{selected: layer.selected}"
```
</blockquote></details>

Done.
___
*[client/web/src/components/panels/LayerTree.vue, line 135 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-Mb_4IN4BCC9R6uvLqQJ:-Mbl9s5z1TyF9-hUcmyl:b-r7bk0y) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/src/components/panels/LayerTree.vue#L135)):*
<details><summary><i>Previously, Keavon (Keavon Chambers) wrote…</i></summary><blockquote>

`j++;`
</blockquote></details>

eslint complains about that, kinda justified
___
*[client/web/src/components/panels/LayerTree.vue, line 210 at r1](https://reviewable.io/reviews/graphiteeditor/graphite/166#-Mb_-syQ1f7yFxY0nCgR:-MblA4-R8HALwWhb2HRO:b-896fix) ([raw file](https://github.com/graphiteeditor/graphite/blob/154a741693476d751661bc925405ed8eb6bd64f3/client/web/src/components/panels/LayerTree.vue#L210)):*
<details><summary><i>Previously, Keavon (Keavon Chambers) wrote…</i></summary><blockquote>

A clearer name might be `selectionRangeStartLayer` and `selectionRangeEndLayer`. It should probably be of type `LayerPanelEntry | undefined` (so in TypeScript I believe this line would say `selectionRangeStartLayer: undefined as LayerPanelEntry | undefined,`). You will store a *reference* to the layer (not only the path) that's also in the `layers` array (see the line above this comment). Then for the range selection, the algorithm can find where the `selectionRangeStartLayer` and `selectionRangeEndLayer` are within the array (and eventually within the tree once we have folders) using  `[a, b, c].indexOf(b)`.

I will leave more detailed feedback on your three click selection functions once you switch over the data type.

Your JavaScript/TypeScript is looking seriously awesome despite your minimal experience using it!
</blockquote></details>

Done.

**Review by Keavon — APPROVED** [review]

**Review by TrueDoctor — COMMENTED** [review]
> Reviewed 2 of 3 files at r3, 1 of 1 files at r4, 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/166)* status: 3 of 4 files reviewed, 8 unresolved discussions (waiting on @akshay1992kalbhor, @Keavon, and @TrueDoctor)

___
*[client/web/src/components/panels/LayerTree.vue, line 175 at r3](https://reviewable.io/reviews/graphiteeditor/graphite/166#-MbC1mHft0NdOvd1yp86-r3-175:-Mbm1cQYB5hn4MrOTgzZ:b3oh8n) ([raw file](https://github.com/graphiteeditor/graphite/blob/3c0c0853401902ebfbb7771e8502db47b2386ba3/client/web/src/components/panels/LayerTree.vue#L175)):*
<details><summary><i>Previously, Keavon (Keavon Chambers) wrote…</i></summary><blockquote>

We apparently need CI to run a newer version of Node.js for this. https://github.com/rustwasm/wasm-bindgen/issues/967
</blockquote></details>


Done.

**Review by TrueDoctor — APPROVED** [review]
> Reviewed 1 of 2 files at r5.
> *[Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/166)* status: all files reviewed, 8 unresolved discussions (waiting on @akshay1992kalbhor and @Keavon)

**Review by Keavon — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/166)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @akshay1992kalbhor)

**Pull request merged**